src/oc_internal_timer.erl

```diff
@@ -0,0 +1,51 @@
+%%%------------------------------------------------------------------------
+%% Copyright 2017, OpenCensus Authors
+%% Licensed under the Apache License, Version 2.0 (the "License");
+%% you may not use this file except in compliance with the License.
+%% You may obtain a copy of the License at
+%%
+%% http://www.apache.org/licenses/LICENSE-2.0
+%%
+%% Unless required by applicable law or agreed to in writing, software
+%% distributed under the License is distributed on an "AS IS" BASIS,
+%% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+%% See the License for the specific language governing permissions and
+%% limitations under the License.
+%%%------------------------------------------------------------------------
+
+-module(oc_internal_timer).
+
+-callback ping() -> ok.
+
+-export([start_link/2,
+         init/1,
+         handle_call/3,
+         handle_cast/2,
+         handle_info/2]).
+
+-record(state, {timer :: reference(),
+                interval :: pos_integer(),
+                module :: module()}).
+
+start_link(Interval, Module) ->
+    gen_server:start_link(?MODULE, {Interval, Module}, []).
+
+init({Interval, Module}) ->
+    Ref = erlang:send_after(Interval, self(), ping),
+
+    {ok, #state{timer = Ref,
+                interval = Interval,
+                module = Module}}.
+
+handle_call(_Msg, _From, State) -> {reply, ok, State}.
+
+handle_cast(_Msg, State) -> {noreply, State}.
+
+handle_info(ping, #state{timer = Ref, interval = Interval, module = Mod}) ->
+    _ = erlang:cancel_timer(Ref),
+    ok = Mod:ping(),
+    NewRef = erlang:send_after(Interval, self(), ping),
```

Comment:
Question is whether we should swap above lines? So if `Mod:ping()` will take longer that `Interval` we should wait another `Interval` or immediately send another event?

Reply:
i like the second...

Reply:
@benoitc you like immediately hitting the event because the send_after is before the call to reporters?

That is what I prefer. I see it as a sort of leak the way it is in this PR. Given a constant rate of spans and the time to report being influenced by the number of spans to report, the time to report will continue to increase and so will the number of spans in the buffer until a potential tipping point.

Reply:
I think that @benoitc meant swapping these two lines, as right now it looks like that (pipe mean interval):

```
  ping 1
|
|
|
  ping 2
```

And with swapped it will be like

```
| ping 1
|
|
  ping 2
```

So this mean that if ping is longer than interval it will end as:

```
| ping 1
| ping 1
  ping 1
  ping 1
  ping 2
```

Instead of

````
  ping 1
  ping 1
  ping 1
|
|
  ping 2
```

Reply:
And if the `ping` call is constantly longer than `Interval` then it will blow up no matter what we will do.

Reply:
Why would that "blow up"?

Reply:
It won't continue to increase like the version that starts the timer after the report.

Reply:
Because there is no way to secure against such problem, at least not when we allow users to configure handlers. Any way that I can came up with will sooner or latter fail in one way or another.

Reply:
Eh, I don't consider it a failure. Though we should definitely update the docs and readme we have now to mention that the time isn't time after finishing reporting but starts before the reporter is called.

Reply:
I can swap them, it is not a big problem. I have written it that way because that was what came to my mind, no hard feelings about it.